pkg/agent/cniserver/ipam/antrea_ipam_controller.go

```diff
@@ -116,35 +116,26 @@ func (c *AntreaIPAMController) Run(stopCh <-chan struct{}) {
 	}()
 
 	klog.InfoS("Starting", "controller", controllerName)
-	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.namespaceInformer.Informer().HasSynced, c.ipPoolInformer.Informer().HasSynced, c.podInformer.HasSynced) {
-		return
+	if c.podInformer != nil && c.namespaceInformer != nil {
+		if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.ipPoolInformer.Informer().HasSynced, c.podInformer.HasSynced, c.namespaceInformer.Informer().HasSynced) {
+			return
+		}
+	} else {
+		if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.ipPoolInformer.Informer().HasSynced) {
+			return
+		}
 	}
 	antreaIPAMDriver.setController(c)
 
 	<-stopCh
 }
 
+// Look up IPPools from the Pod annotation.
 func (c *AntreaIPAMController) getIPPoolsByPod(namespace, name string) ([]string, []net.IP, *crdv1a2.IPAddressOwner, error) {
-	// Find IPPool by Pod
 	var ips []net.IP
 	var reservedOwner *crdv1a2.IPAddressOwner
 	pod, err := c.podLister.Pods(namespace).Get(name)
 	if err != nil {
-		// For CNI DEL case, getting Pod may fail. Try to get information from allocated IPs of all IPPools.
```

Comment:
This logic is removed in the new code, is it possible to lead to IP leak in CNI DEL event because IPPool fails to get as Pod is removed?

Reply:
Checked Jianjun's change, I think it will use a new function to check ipPool when CNIDel, thus this is OK.
`func (d *AntreaIPAM) del(podOwner *crdv1a2.PodOwner) (foundAllocation bool, err error)`

Reply:
Yes, I changed the code to always look up IPPools by Pod. I think it might be possible a pool status update is not received by agent yet when CNI DEL comes, and so we might not be able to release the allocated IP (@gran-vmv : do you agree?). But this can happen even with the existing code.

I made change more for secondary network and primary network to share the same DEL code. Also, I feel in this way the code for ADD is also simpler.

Reply:
@jianjuns I think your method should work, but not sure if the cost for searching PodOwner in IPPools for all PodDel is acceptable.

Reply:
But the cache is indexed by Pod name (+ Namespace)? Beside code sharing, I also feel this approach is safer - it can cover Pod annotation or CNI config change (for secondary network).